### PR TITLE
Added 'duplicate' function for Dictionary in C++ and API.

### PR DESCRIPTION
**Pull request by willnationsdev**
Now we don't have to do `var2str(str2var(dict))` anymore~

### CODE REVIEW

**Comment by hpvb** [comment]
LGTM maybe port this to 2.1 also?

**Comment by willnationsdev** [comment]
@hpvb Can I even do that in the same PR?

**Comment by hpvb** [comment]
@willnationsdev no you can't, I was hoping you'd make another PR for the 2.1 branch :)

File: core/variant_call.cpp [hunk 1]
```diff
@@ -1607,6 +1608,8 @@ void register_variant_methods() {
 	ADDFUNC0R(DICTIONARY, INT, Dictionary, hash, varray());
 	ADDFUNC0R(DICTIONARY, ARRAY, Dictionary, keys, varray());
 	ADDFUNC0R(DICTIONARY, ARRAY, Dictionary, values, varray());
+	//renaming "copy" to "duplicate" to mimic Array and Node method APIs
```
Comment (akien-mga) [inline core/variant_call.cpp]: How often is `Dictionary.copy()` used through the codebase? Might be worth renaming to `duplicate` to be consistent.

Comment (Hinsbart) [reply core/variant_call.cpp]: > How often is Dictionary.copy() used through the codebase?

If QtCreators reference search is to be trusted, exactly 10 times. ^^

Comment (akien-mga) [reply core/variant_call.cpp]: Then I'd say go for it.

**Comment by blurymind** [comment]
It would help if str2var was a bit more flexible too
https://github.com/godotengine/godot/issues/13021

**Comment by Zylann** [comment]
Does Array have duplicate too? How does it deal with object, nested dictionaries, arrays and cycles? Is it deep or shallow copy?

**Comment by Zylann** [comment]
Oh wait... It existed already? (duh cant edit on phone)

**Comment by willnationsdev** [comment]
Seems like it does a shallow copy, just directly copying over whatever it is.

**Comment by reduz** [comment]
I see Array has duplicate and Dictionary has copy().
I think this PR can be modified to:
1) Make both have the same name
2) Add an option for recursive deep copy in both, with default argument as false

Would this be good?

**Comment by willnationsdev** [comment]
@reduz I'm on it.

**Comment by Zylann** [comment]
What would recursive copy do though? How would it handle resources, nodes, objects, cycles etc?

**Comment by willnationsdev** [comment]
Well, resources don't conceptually make sense for "deep copying", so I would assume the behavior for them remains the same. For arrays and dictionaries, I can just use the copy constructor I'm pretty sure. For objects, we'd need to be able to get a new instance of whatever type of object it is with all of the properties/signals, etc. copied over. For nodes, you'd just call their duplicate function (although that one takes flags because it's so complicated...). I suppose objects/nodes would be more complicated to handle...

**Comment by willnationsdev** [comment]
rather than a simple boolean for deep copying, you would need to have flags passed in, the same as node's duplicate method

**Comment by willnationsdev** [comment]
Hmmm, this might amount to something that needs a lot of testing. Maybe for now we just keep this PR as a duplicate renaming and then open a separate Issue to discuss the options for deep copying things?

**Comment by willnationsdev** [comment]
This is with the renaming of copy to duplicate made across the engine. (I grepped for every mention I could find related to the `Dictionary` class).

**Comment by reduz** [comment]
> What would recursive copy do though? How would it handle resources, nodes, objects, cycles etc?

No, deep copy is just for arrays and dictionaries. It should not apply to objects (would be very unintuitive). Regarding cycles, you can put a recursion limit of say 1024 and throw an error if it's reached.

**Pull request merged**